### ISSUE TITLE
bugged pads on savefig

### ISSUE THREAD

**Issue by IvanAvdeevGth** [matplotlib/matplotlib#18804]
Found that pads do not work correctly. Previously (in approximately 2016) the code below would save a black square, but nowadays it saves a square with tiny white strips (around the square). Not a big deal, but it seems that it is not supposed to work in that way.

```python
#!/usr/bin/env python3
import matplotlib.pyplot as plt
fig = plt.figure(0, figsize=(4,4))
ax = fig.add_axes((0,0,1,1))
ax.set_xticklabels([])
ax.set_yticklabels([])
ax.set_xticks([])
ax.set_yticks([])
ax.set_facecolor((0,0,0))
fig.savefig('foo.pdf', bbox_inches='tight', pad_inches=0)

Python version 3.6.9
matplotlib version 3.3.2
```

**Comment by jklymak** [comment]
Yes, it looks like there is a stray tick at 0 on both the x and y axis.  Not sure yet what causes this.

**Comment by jklymak** [comment]
This has been like this since at least 2.0.  OTOH I don't understand why we have a spare visible tick lying around.

**Comment by jklymak** [comment]
bumping forward - someone should look at this, but given its long-standing presence, perhaps not urgent.

**Comment by anntzer** [comment]
Looks like the following fully fixes this issue and helps with #16981 (which makes Axis invisible rather than disabling ticks), although some residual strip remains for #16981:
```patch
diff --git i/lib/matplotlib/spines.py w/lib/matplotlib/spines.py
index 40ceb157f..28ea56733 100644
--- i/lib/matplotlib/spines.py
+++ w/lib/matplotlib/spines.py
@@ -151,15 +151,19 @@ class Spine(mpatches.Patch):
         # make sure the location is updated so that transforms etc are correct:
         self._adjust_location()
         bb = super().get_window_extent(renderer=renderer)
-        if self.axis is None:
+        if self.axis is None or not self.axis.get_visible():
             return bb
         bboxes = [bb]
-        tickstocheck = [self.axis.majorTicks[0]]
-        if len(self.axis.minorTicks) > 1:
-            # only pad for minor ticks if there are more than one
-            # of them.  There is always one...
-            tickstocheck.append(self.axis.minorTicks[1])
-        for tick in tickstocheck:
+        drawn_major_tick = None
+        drawn_minor_tick = None
+        for tick in self.axis._update_ticks():
+            if drawn_major_tick is None and tick in self.axis.majorTicks:
+                drawn_major_tick = tick
+            elif drawn_minor_tick is None and tick in self.axis.minorTicks:
+                drawn_minor_tick = tick
+        for tick in [drawn_major_tick, drawn_minor_tick]:
+            if tick is None:
+                continue
             bb0 = bb.frozen()
             tickl = tick._size
             tickdir = tick._tickdir
```
The point is that majorTicks[0] is not necessarily being drawn; you need to call `_update_ticks` to figure out what's actually getting drawn -- if anything.

Come claim your free PR :)
